bdd/spec/017_custom_event_lnn_spec.sh

```diff
@@ -0,0 +1,148 @@
+Include build_tools.sh
+
+Describe "Custom events"
+  Describe "loop custom event"
+    before() {
+      sourceToTemp "
+        from @std/app import start, print, exit
+
+        event loop: int64
+
+        on loop fn looper(val: int64) {
+          print(val);
+          if val >= 10 {
+            emit exit 0;
+          } else {
+            emit loop val + 1 || 0;
+          }
+        }
+
+        on start {
+          emit loop 0;
+        }
```

Comment:
I am a little hesitant on replicating this test as-is. It is easy to produce an infinite loop with this pattern, so I was hoping to block it in a similar way the direct recursive function calling is detected and blocked.

Reply:
but infinite loops using this pattern is safe right? It doesn't have the same issue as infinite loops because it doesn't block the runtime and it doesn't have the same issue as infinite recursion because it creates a new task for the event handler?

Reply:
That's why I first allowed it, but I'm still ambivalent about it. If you're trying to do a `setInterval` type thing, it would probably be better for us to add a language feature to properly expose `setInterval` and the Tokio equivalent instead of a quasi-hot-loop polling the current time and seeing if it should run or not?

Basically, it is similar to a `while` loop in that providing functions or language-level constructs to directly tackle the patterns the user wants instead gives the compiler and runtime the ability to be more efficient with what they're doing. This test was very much just a convenient way to test event emission from conditionals, but actually needing this pattern feels more like we're missing a better abstraction in the language that we should add, instead?

Reply:
Honestly, I think that this pattern for `setInterval` is the most intuitive for Alan since it's event-based. We *could* expose *yet more* opcodes for this, but this feels like a really natural pattern for event-based languages and should probably be considered idiomatic over eg `@std/seq`.

Reply:
So I wasn't prescribing how the event registration should look (I don't think it should look like JS's `setInterval`, I was mostly pointing out that using a loop like this would either: 1) be very inefficient being a quasi-hot-loop or 2) would be very loose on when it actually runs by needing a `wait` call inside of the event-based loop.

Having this construct at the language level would probably be a safer bet, *especially* in the cluster-of-AVMs perspective when you might want the interval to only run *once* for the entire cluster every N minutes, while using this loop is going to have it run M times for the number of nodes in the cluster.

Reply:
> 1) be very inefficient being a quasi-hot-loop or 2) would be very loose on when it actually runs by needing a wait call inside of the event-based loop.

While true, I don't think that's enough of a reason to prevent this pattern. There might be scenarios where a function needs to run *roughly* once every hour - this pattern would be perfect for that.

Or, if a user is trying to implement a state machine, this pattern would be quite handy. If we block access to this pattern, the user would have to use `@std/seq`, which intentionally requires loops to be hard-capped at a specific number. That means that a user writing a state machine that should be running for a long time will have to rely on a stable number of http requests *or* will have to restart their program at the same interval. This allows the user to write an infinite loop if they need it, in a safe and idiomatic manner.

> especially in the cluster-of-AVMs perspective when you might want the interval to only run once for the entire cluster every N minutes, while using this loop is going to have it run M times for the number of nodes in the cluster.

This would definitely be very useful and I think we should include it as well, but that feels like a different use-case to me

Reply:
I think it would be the standard use case,  though? When the cluster autoscalers to who knows how many nodes, making sure something is run on each node is going to be far less common of a need than wanting the event to happen once at a fixed interval.

But I can see value in leaving the other form open for the rarer cases that it is desired. (Particularly around monitoring.)